source/simulator/parameters.cc

```diff
@@ -1124,6 +1146,124 @@ namespace aspect
+        {
+          // each entry has the format (white space is optional):
+          // <name> : <value (might have spaces)> [component]
+          //
+          // first tease apart the two halves
+          const std::vector<std::string> split_parts = Utilities::split_string_list (*p, ':');
+          AssertThrow (split_parts.size() == 2,
+                       ExcMessage ("The format for mapped particle properties  "
+                                   "requires that each entry has the form `"
+                                   "<name of field> : <particle property> [component]', "
+                                   "but there does not appear to be a colon in the entry <"
+                                   + *p
+                                   + ">."));
+
+          // the easy part: get the name of the compositional field
+          const std::string key = split_parts[0];
+
+          // check that the names used are actually names of fields,
+          // are solved by particles, and are unique in this list
+          std::vector<std::string>::iterator field_name_iterator = std::find(names_of_compositional_fields.begin(),
+                                                                             names_of_compositional_fields.end(), key);
+          AssertThrow (field_name_iterator
+                       != names_of_compositional_fields.end(),
+                       ExcMessage ("Name of field <" + key +
+                                   "> appears in the list of mapped particle properties, but"
+                                   "there is no field with this name."));
+
+          const unsigned int compositional_field_index = std::distance(names_of_compositional_fields.begin(),field_name_iterator);
+          AssertThrow (compositional_field_methods[compositional_field_index]
+                       == "particles",
+                       ExcMessage ("The field <" + key +
+                                   "> appears in the list of mapped particle properties, but"
+                                   "is not advected by a particle method."));
+
+          AssertThrow (std::count(names_of_compositional_fields.begin(),
+                                  names_of_compositional_fields.end(), key) == 1,
+                       ExcMessage ("Name of field <" + key +
+                                   "> appears more than once in the list of "
+                                   "mapped particle properties."));
+
+          // now for the rest. since we don't know whether there is a
+          // component selector, start reading at the end and subtract
+          // a number that might be a component selector
+          std::string value_and_comp = split_parts[1];
+          std::string comp;
+          if ((value_and_comp.size()>0) &&
+              ((value_and_comp[value_and_comp.size()-1] >= '0')
+               &&
+               (value_and_comp[value_and_comp.size()-1] <= '9')))
```

Comment:
I had assumed from the description in the `declare_parameter` call that one would append the component selector in square brackets, in the spirit of an indexing operation. I realize now that you did not mean the square brackets that way, but as optional. Either way, it might be easier to actually *use* square brackets to select a component, because it would make parsing so much simpler. Want to consider?